LayoutTests/media/media-source/media-source-append-buffer-full-evict-prior-to-end.html

```diff
@@ -0,0 +1,87 @@
+<!DOCTYPE html>
+<html>
+<head>
+    <title>mock-media-source</title>
+    <script src="mock-media-source.js"></script>
+    <script src="../video-test.js"></script>
+    <script>
+    var source;
+    var sourceBuffer;
+    var initSegment;
+    var exception;
+
+    function bufferedRanges() {
+        var bufferedRanges = '[ ';
+        var timeRanges = sourceBuffer.buffered;
+        for (var i = 0 ; i < timeRanges.length ; i++) {
+            if (i)
+                bufferedRanges += ', ';
+            bufferedRanges += timeRanges.start(i) + '...' + timeRanges.end(i);
+        }
+        bufferedRanges += ' ]';
+        return bufferedRanges;
+    }
+
+    function doBufferedRangesContain(time) {
+        var timeRanges = sourceBuffer.buffered;
+        for (var i = 0 ; i < timeRanges.length ; i++) {
+            if (timeRanges.start(i) <= time && time <= timeRanges.end(i))
+                return true;
+        }
+        return false;
+    }
+
+    async function appendPtsRange(firstPts, lastPts) {
+        var resultException = null;
+        for (var pts = firstPts; pts <= lastPts; pts++) {
+            try {
+                consoleWrite('Appending PTS='+pts);
+                sourceBuffer.appendBuffer(makeASample(pts, pts, 1, 1, 1, pts % 5 == 0 ? SAMPLE_FLAG.SYNC : SAMPLE_FLAG.NONE, 1));
```

Comment:
This is a nice simple test, but it would also be nice to test our code doesn't break in presence of B-frames. Here you have some example frames with B-frames:
![b-frames](https://user-images.githubusercontent.com/1002436/201664392-bf525a0d-c74f-4304-8324-50cc82e25a41.jpg)

Reply:
Addedum: in this example I've given the second I-frame is not an IDR-frame, and therefore should not be considered a sync sample as far as WebKit is concerned. [Quoting the MSE spec](https://www.w3.org/TR/media-source/#random-access-point):

> Random Access Point: A position in a [media segment](https://www.w3.org/TR/media-source/#media-segment) where decoding ***and continuous playback*** can begin without relying on any previous data in the segment. For video this ***tends*** to be the location of I-frames.

The second I-frame is not a random access point/sync sample/IDR frame because, although it is enough to decode anything coming after it in decode order, it's not enough to play everything after it in presentation order. Notably, the B-frame that is immediately after it in presentation order cannot be played without the previous GOP.